File: framework/play/src/main/scala/play/api/mvc/Security.scala
package play.api.mvc

import play.api._
import play.api.mvc.Results._

/**
 * Helpers to create secure actions
 */
object Security {

  /**
   * Key of the USERNAME attribute stored in session.
   */
  val USERNAME = "username"

  /**
   * Wrap another action, allowing only authenticated HTTP requests.
   *
   * The username is retrieved from the session cookie, and added to the HTTP request
   * username attribute.
   *
   * Example:
   * {{{
   * Authenticated {
   *   Action { request =>
   *     Ok(request.username.map("Hello " + _))
   *   }
   * }
   * }}}
   *
   * @tparam A Type of the request body.
   * @param username Function used to retrieve the username from the request header (default is to read from session cookie).
   * @param onUnauthorized Function used to generate alternative result if the user is not authenticated (default to simple 401 page)
   * @param action Action to wrap.
   */
  def Authenticated[A](
    username: RequestHeader => Option[String] = (req => req.session.get(USERNAME)),
    onUnauthorized: RequestHeader => Result = (_ => Unauthorized(views.html.defaultpages.unauthorized())))(action: Action[A]): Action[A] = {

    action.compose { (request, originalAction) =>
      username(request).map { authenticatedUser =>
        val authenticatedRequest = new Request[A] {
          def uri = request.uri
          def path = request.path
          def method = request.method
          def queryString = request.queryString
          def headers = request.headers
          def cookies = request.cookies
          def username = Some(authenticatedUser)
          val body = request.body
        }
        originalAction(authenticatedRequest)
      }.getOrElse {
        onUnauthorized(request)
      }
    }

  }

  /**
   * Wrap another action, allowing only authenticated HTTP requests.
   *
   * The username is retrieved from the session cookie, and added to the HTTP request
   * username attribute.
   *
   * Example:
   * {{{
   * Authenticated {
   *   Action { request =>
   *     Ok(request.username.map("Hello " + _))
   *   }
   * }
   * }}}
   *
   * @tparam A Type of the request body.
   * @param action Action to wrap.
   */
  def Authenticated[A](action: Action[A]): Action[A] = Authenticated()(action)

  /**
   * Redefines the default Controller Action method, to make all actions authenticated.
   *
   * Example:
   * {{{
   * object Admin extends Controller with AllAuthenticated {
   *
   *   def index = Action { request =>
   *     Ok(request.username.map("Hello " + _))
   *   }
   *
   * }
   * }}}
   */
  trait AllAuthenticated extends ControllerLike {

    /**
     * Override the default Action method.
     *
     * All Action defined in this controller will be wrapped in an Authenticated action.
     *
     * @see Authenticated
     */
    override abstract def Action[A](bodyParser: BodyParser[A], block: Request[A] => Result): Action[A] = {
      Authenticated(
        onUnauthorized = (req => onUnauthorized(req)),
        username = (req => username(req)))(super.Action(bodyParser, block))
    }

    /**
     * Retrieve the username from the request header (default is to read from session cookie).
     *
     * @param request The request header.
     * @return Maybe the username.
     */
    def username(request: RequestHeader): Option[String] = request.session.get("username")

    /**
     * Generate alternative result if the user is not authenticated (default to simple 401 page)
     *
     * @param request The request header.
     * @return The result to be sent to the user if the request was not authenticated.
     */
    def onUnauthorized(request: RequestHeader): Result = Unauthorized(views.html.defaultpages.unauthorized())

  }

}

File: framework/src/play/src/main/java/play/mvc/Security.java
package play.mvc;

import play.libs.F;
import play.mvc.Http.*;

import java.lang.annotation.*;

/**
 * Defines several security helpers.
 */
public class Security {
    
    /**
     * Wraps the annotated action in an <code>AuthenticatedAction</code>.
     */
    @With(AuthenticatedAction.class)
    @Target({ElementType.TYPE, ElementType.METHOD})
    @Retention(RetentionPolicy.RUNTIME)
    public @interface Authenticated {
        Class<? extends Authenticator> value() default Authenticator.class;
    }
    
    /**
     * Wraps another action, allowing only authenticated HTTP requests.
     * <p>
     * The user name is retrieved from the session cookie, and added to the HTTP request's
     * <code>username</code> attribute.
     */
    public static class AuthenticatedAction extends Action<Authenticated> {
        
        public F.Promise<SimpleResult> call(Context ctx) {
            try {
                Authenticator authenticator = configuration.value().newInstance();
                String username = authenticator.getUsername(ctx);
                if(username == null) {
                    Result unauthorized = authenticator.onUnauthorized(ctx);
                    if (unauthorized instanceof AsyncResult) {
                        return ((AsyncResult) unauthorized).getPromise();
                    } else {
                        return F.Promise.pure((SimpleResult) unauthorized);
                    }
                } else {
                    try {
                        ctx.request().setUsername(username);
                        return delegate.call(ctx);
                    } finally {
                        ctx.request().setUsername(null);
                    }
                }
            } catch(RuntimeException e) {
                throw e;
            } catch(Throwable t) {
                throw new RuntimeException(t);
            }
        }

    }
    
    /**
     * Handles authentication.
     */
    public static class Authenticator extends Results {
        
        /**
         * Retrieves the username from the HTTP context; the default is to read from the session cookie.
         *
         * @return null if the user is not authenticated.
         */
        public String getUsername(Context ctx) {
            return ctx.session().get("username");
        }
        
        /**
         * Generates an alternative result if the user is not authenticated; the default a simple '401 Not Authorized' page.
         */
        public Result onUnauthorized(Context ctx) {
            return unauthorized(views.html.defaultpages.unauthorized.render());
        }
        
    }
    
    
}
